Allow `--update-missing` to change the default `how` value

happz opened the pull request:
`--update-missing` did not touch `how`. However, when user did not specify `how` in a phase, they can reasonably expect `--update-missing` would set it. Unfortunately, tmt first resolves default values, and sets the `how` key to the default value, which makes it no longer missing when CLI invocations are being resolved.

The patch saves original, fmf-supplied `how` and `name`, and `--update-missing` checks whether `how` has been unset - if that's the case, `how` would be update if `--update-missing --how ...` is given by user.

Pull Request Checklist

* [x] implement the feature
* [x] write the documentation
* [x] extend the test coverage

happz commented:
> Some tests seem to be failing.

Ack. I'll look into them in the evening/tomorrow, but it shouldn't affect whatever @janhavlin might be interested in testing for the multihost pipeline.

happz commented:
FTR, I'm fairly unhappy about how I solved the problem, and I'm looking into other possibilities :/

happz commented:
@psss I moved the code propagating `--insert`, `--update` and `--update-missing` to part of the code before `Step.wake()` is called. it was done in `wake()`, mostly because the impression was this is where it has to happen to support re-running tmt with existing workdir, but with different CLI options. now when I moved it to happen sooner, I think this particular functionality is probably harmed and not supported anymore - but I couldn't find any test that would fail because of this. the only mention of "wake" is in `/tests/core/force`, and that test passes with my changes.

psss commented:
> I think this particular functionality is probably harmed and not supported anymore - but I couldn't find any test that would fail because of this.

Right you are, the expected behavior of command line options overriding the woken `step.yaml` data is broken by this change. This really has not been covered by any test yet. And it's possible to reproduce only when the new `--again` option is used. In 3682070e I've added a new test to the #2528 pull request which covers this. You can use it to verify, that the `--update-missing` change will work for this use case as well.

happz commented:
> > I think this particular functionality is probably harmed and not supported anymore - but I couldn't find any test that would fail because of this.
> 
> Right you are, the expected behavior of command line options overriding the woken `step.yaml` data is broken by this change. This really has not been covered by any test yet. And it's possible to reproduce only when the new `--again` option is used. In [3682070](https://github.com/teemtee/tmt/commit/3682070e4aecbac471adb4679211644b1130a486) I've added a new test to the #2528 pull request which covers this. You can use it to verify, that the `--update-missing` change will work for this use case as well.

Fixed my PR, now it should pass all tests including the new one.

thrix commented:
/packit test --identifier full

thrix commented on tmt/base.py:
```diff
@@ -1567,6 +1575,11 @@ def __init__(
             raise_on_validation_error=raise_on_validation_error,
             **kwargs)
 
+        # TODO: there is a bug in handling internal fields with `default_factory`
```
@happz this is planned to be addressed somewhere?

happz replied:
Filed and issue, referenced in https://github.com/teemtee/tmt/issues/2630

thrix reviewed (approved):
LGTM, no more comments found

psss reviewed (changes requested):
Thanks much for fixing this! Looks good, just a few minor comments.

psss commented:
/packit test -i full

happz commented on tests/core/phases/test.sh:
```diff
@@ -160,6 +160,42 @@ default-1:html:50"
+        check "with-report" "One 'html' phase shall exist" "default-0:html:50"
+
+        rlRun -s "yq -r '.data | .[] | \"\\(.name):\\(.how)\"' $rundir/plans/with-report/report/step.yaml"
+        rlAssertEquals "default-0 shall be set to html how" "$(cat $rlRun_LOG)" "default-0:html"
```
Hmm, it does seem weird. I've found a couple more similar asserts, I'd say this needs to be reviewed. Preferably in another patch, I'm not sure I want to break them all now :)

psss replied:
Ok, filed as #2638.

happz commented on tmt/base.py:
```diff
@@ -1539,6 +1539,14 @@ class Plan(
     _original_plan: Optional['Plan'] = field(default=None, internal=True)
     _remote_plan_fmf_id: Optional[FmfId] = field(default=None, internal=True)
 
+    #: Used by steps to mark invocations that has been already applied to
```
Fixed in 84de32bd

happz commented on tmt/steps/__init__.py:
```diff
@@ -369,44 +374,80 @@ def __init__(
 
-        for _i, raw_datum in enumerate(raw_data):
+        for i, raw_datum in enumerate(raw_data):
+            debug1(f'raw step datum #{i} in ', str(raw_datum), shift=1)
```
Fixed in 84de32bd

happz commented on tmt/steps/__init__.py:
```diff
@@ -369,44 +374,80 @@ def __init__(
 
+                debug2(f"setting 'name' to default '{raw_datum['name']}'", shift=2)
+
+            debug1(f'raw step datum #{i} out', str(raw_datum), shift=1)
```
Fixed in 84de32bd

happz commented on tmt/steps/__init__.py:
```diff
@@ -704,15 +787,17 @@ def _patch_raw_datum(
                     continue
 
+                # The `how` has not been set in fmf, therefore --update-missing
+                # applies to `how` as well.
```
Fixed in 84de32bd

psss commented:
/packit test -i full
